### ISSUE TITLE
Search on a_pp

### ISSUE THREAD

**Issue by danyasaki** [LMFDB/lmfdb#2134]
It would be nice to be able to search on the a_pp for various objects.  When the a_pp are in QQ, this might be achieved by storing the pp_label: a_pp pair in mongodb.  This may involve more trickery when the a_pp is not in QQ.

**Comment by jvoight** [comment]
I guarantee we will receive these requests--and the same for the Atkin-Lehner eigenvalues.  

Maybe we search only over the first many (not all)?  If we have many eigenvalues, I'm not sure that listing them each as a field is a good idea...  Isn't this what is done for classical modular forms?  We could add fields for the first few eigenvalues to search on them in addition to the list as a string (or an array of strings).

We could list the minimal polynomial of the eigenvalue when it's not over QQ, or the trace.  Or actually, we could also store the Hecke field via its LMFDB label when that can be computed--it's probably the most convenient way to search for it!  (I realize we this would require extra computation, and wouldn't be available for all forms, but is probably the way users would want to interact.)

**Comment by JohnCremona** [comment]
We'll need to think carefully about exactly what we store and in what format.  A simple list (whether of integers or strings) is not so good for searching -- even if you know that your prime P of interest is (say) the 3rd prime I don't know if it is possible (or efficient if possible) to search for records where the 3rd entry in some list has some value.  A dictionary would work (see https://stackoverflow.com/questions/7431422/find-in-dictionary-by-value-in-mongo for how).
In elliptic_curves.curves I store lists of a_n and a_p as lists, and realise now that was a mistake since searching is hard.

**Comment by jwj61** [comment]
It is possible to search on the 3rd entry in a list.  Harold told me once, but it was some years ago and I don't remember the syntax.

**Comment by AndrewVSutherland** [comment]
@jwj61 Yes this is possible, for example if db is set to localfields in python you could use:

> db.fields.find_one({'coeffs.2':int(17)})

to find a local field in which the third element of the coeffs field is 17 (array indexing starts at 0).

But you cannot directly access/search elements in nested lists.

**Comment by JohnCremona** [comment]
OK, so the syntax for the 2nd element of a list is the same as for a dictionary value with '2' as key.  I wonder which is more efficient.

**Comment by jvoight** [comment]
OK, currently the Hecke eigenvalues for HMFs are stored as a string, so to satisfy this, we should be storing these as a list?  Is that what is done for BMFs and classical modular forms?

**Comment by JohnCremona** [comment]
The eigenvalues are now stored as a list of strings, not as a single string.  I followed this example for Bianchi forms, with the additional feature that for dimension 1 when they are integers, I store them as integers.  (If that is a bad idea I can change it.)
Classical modular forms?  I don't know, as these are not visible via the api.

**Comment by AndrewVSutherland** [comment]
The trace table search at http://www.lmfdb.org/ModularForm/GL2/Q/holomorphic/?search_type=Traces&all=1 provides this functionality for classical modular forms when a_pp is in Q.
